# Introduction
[Reference](https://www.hackerrank.com/challenges/summing-the-n-series/problem?isFullScreen=true)
- There is a sequence whose $n^{\text{th}}$ term is $$T_n = n^2 - (n-1)^2$$
- Evaluate the series
$$ S_n = \sum_{k=1}^n T_k = T_1 + T_2 + \ldots + T_n $$
Find $S_n$ mod $(10^9 + 7)$ (for truncating while $n$ is large enough).

#### Example
For $n=3$ Then the series is $(1 - 0) + (2^2 - 1) + (3^2 - 2^2) = 9$.

#### Function Description
Complete the `summingSeries function` with the following parameter(s):
- int n: the inclusive limit of the range to sum

#### Returns
- int: the sum of the sequence, modulo 

#### Input Format
- The first line of input contains $t$, the number of test cases.
- Each test case consists of one line containing a single integer .

#### Constraints
- $1 \leq t \leq 10$
- $1 \leq n \leq 10^{16}$


In [1]:
def summingSeries(n):
    # Write your code here
    return n**2 %(10**9 + 7)

check_list = [4, 102, 1010, 121244, 288121244, 14230061991]
for num in check_list:
    print(f"Summing of n({num})-series is {summingSeries(num)}")

Summing of n(4)-series is 16
Summing of n(102)-series is 10404
Summing of n(1010)-series is 1020100
Summing of n(121244)-series is 700107438
Summing of n(288121244)-series is 663010586
Summing of n(14230061991)-series is 240244131


## 2. Summing of square number
What happend if 

$$ S_n = \sum_{k=1}^n k^2 = 1 + 2^2 + \ldots + n^2 $$

In [2]:
def squared_summing_loop(n):
    Sum = 0
    for k in range(1, n+1):
        Sum += k**2
    return Sum % (10**9 + 7)

check_list = [4, 69, 102, 1010, 12121, 121244, 19912100]
for num in check_list:
    print(f"Summing of n({num})-series is {squared_summing_loop(num)}")

Summing of n(4)-series is 30
Summing of n(69)-series is 111895
Summing of n(102)-series is 358955
Summing of n(1010)-series is 343943885
Summing of n(12121)-series is 673739710
Summing of n(121244)-series is 291946821
Summing of n(19912100)-series is 92152201


##### Disadvantage
When $n$ is large enough, we must apply another method, knowing that
$$ S_n = \sum_{k=1}^n k^2 = \frac{n(n+1)(2n+1)}{6} $$

In [3]:
def squared_summing_generalized(n):
    Sum = n*(n + 1)*(2*n + 1) // 6
    return Sum % (10**9 + 7)

check_list = [4, 102, 12121, 121244, 19912100, 288121244]
for num in check_list:
    print(f"Summing of n({num})-series is {squared_summing_generalized(num)}")

Summing of n(4)-series is 30
Summing of n(102)-series is 358955
Summing of n(12121)-series is 673739710
Summing of n(121244)-series is 291946821
Summing of n(19912100)-series is 92152201
Summing of n(288121244)-series is 208291149


## 3. Suumming of cube
Likewise, what is the output of 

$$ S_n = \sum_{k=1}^n k^3 = 1 + 2^3 + \ldots + n^3 $$

In [4]:
def cubed_sum_loop(n):
    Sum = 0
    for k in range(1, n+1):
        Sum += k**3
    return Sum % (10**9 + 7)

check_list = [4, 69, 102, 1010, 12121, 121244, 19912100]
for num in check_list:
    print(f"Summing of n({num})-series is {cubed_sum_loop(num)}")

Summing of n(4)-series is 100
Summing of n(69)-series is 5832225
Summing of n(102)-series is 27594009
Summing of n(1010)-series is 666406205
Summing of n(12121)-series is 167695027
Summing of n(121244)-series is 915803931
Summing of n(19912100)-series is 297101031


Also, we have

$$ S_n = \sum_{k=1}^n k^3 = \frac{n^2(n+1)^2}{4} $$

In [5]:
def cubed_sum(n): return (n**2 * (n+1)**2 // 4)  % (10**9 + 7)
check_list = [4, 102, 12121, 121244, 19912100, 288121244]
for num in check_list:
    print(f"Summing of n({num})-series is {cubed_sum(num)}")

Summing of n(4)-series is 100
Summing of n(102)-series is 27594009
Summing of n(12121)-series is 167695027
Summing of n(121244)-series is 915803931
Summing of n(19912100)-series is 297101031
Summing of n(288121244)-series is 926646666


## 4. Summary
- We have
$$ \sum_{k=1}^n k^3 = \frac{n^2(n+1)^2}{4} = \left( \frac{n(n+1)}{2} \right)^2 = \left( \sum_{k=1}^n k \right)^2 $$

So, what is the strategy to compute $ \displaystyle S_{n, m} = \sum_{k=1}^n k^m$, for instance $m = 4$??

- Noting that for any degree $m$ in the left hand side we will obtain a polynomial with $m+1$ degrees in the right hand side

$$ a_{m+1} n^{m+1} + a_m n^m + a_{m-1} n^{m-1} + \ldots + a_0 $$

- For $n=0$ then this is easy to get

$$ S_{0, m} = \sum_{k=0}^{m+1} a_k 0^k \Leftrightarrow a_0 = 0$$

- Hence, to determine $a_1, \ldots, a_{m+1}$

In [6]:
m = 5
for n in range(1, m+1):
    S1 = 0
    for k in range(1, n+1):
        S1 += k**4
    print(f"S({n}) = {S1}")

S(1) = 1
S(2) = 17
S(3) = 98
S(4) = 354
S(5) = 979


Then, we have:
- $S_1$ implies $a_1 + a_2 + a_3 + a_4 + a_5 = 1$
- $S_2$ implies $a_1 * 2 + \ldots + a_5 * 2^5= 17$ or $2a_1 + 4a_2 + 8a_3 + 16 a_4 + 32a_5= 17$
- $S_3$ implies $3a_1 + 9a_2 + 27a_3 + 81a_4 + 243a_5= 98$
- $S_4$ implies $4a_1 + 16a_2 + 64a_3 + 256a_4 + 1024a_5= 354$
- $S_5$ implies $5a_1 + 25a_2 + 125a_3 + 625a_4 + 3125a_5= 979$

and we obtain

$$ \left( \begin{array}{cccc} 1 & 1 & 1 & 1 & 1\\ 2 & 4 & 8 & 16 & 32 \\ 3 & 9 & 27 & 81 & 243 \\ 4 & 16 & 64 & 256 & 1024 \\ 5 & 25 & 125 & 625 & 3125 \end{array} \right) \left( \begin{array}{c} a_1 \\ a_2 \\ a_3 \\ a_4 \\ a_5 \end{array} \right) = \left( \begin{array}{c} 1 \\ 17 \\ 98 \\ 354 \\ 979 \end{array} \right) $$

In [7]:
import numpy as np

m = 4
def sum_power_coef(m, is_print=False):
    def sum_power(m):
        S_list = []
        for n in range(1, m+1):
            S1 = 0
            for k in range(1, n+1):
                S1 += k**(m-1)
            S_list.append(S1)
        return S_list

    A = np.array([[k**d for d in range(1, m+2)] for k in range(1, m+2)])
    B = sum_power(m + 1)
    if is_print:
        print(A, B)
    else:
        pass
    return np.matmul(np.linalg.inv(A), B) 

def sum_power_series(m, n, is_print=False):
    coef_ = sum_power_coef(m, is_print)
    if is_print:
        print("coef:", coef_)
    else:
        pass
    Sum = 0
    for k in range(0, m+1):
        Sum += coef_[k] * n**(k+1) 
    return int(np.ceil(Sum)) % (10**9 + 7)

sum_power_series(4, 20, True)

[[   1    1    1    1    1]
 [   2    4    8   16   32]
 [   3    9   27   81  243]
 [   4   16   64  256 1024]
 [   5   25  125  625 3125]] [1, 17, 98, 354, 979]
coef: [-3.33333333e-02 -1.75925940e-12  3.33333333e-01  5.00000000e-01
  2.00000000e-01]


722667

In [8]:
sum_power_series(1, 20, True)

[[1 1]
 [2 4]] [1, 3]
coef: [0.5 0.5]


210

### Verification!!

In [9]:
for num in [2,4,6,10,200,5000]:
    print(f"Summing of n({num})-series by func `cubed_sum` is {cubed_sum(num)} \n\t while generalized-func is {sum_power_series(3, num)}")

Summing of n(2)-series by func `cubed_sum` is 9 
	 while generalized-func is 9
Summing of n(4)-series by func `cubed_sum` is 100 
	 while generalized-func is 100
Summing of n(6)-series by func `cubed_sum` is 441 
	 while generalized-func is 441
Summing of n(10)-series by func `cubed_sum` is 3025 
	 while generalized-func is 3025
Summing of n(200)-series by func `cubed_sum` is 404010000 
	 while generalized-func is 404010000
Summing of n(5000)-series by func `cubed_sum` is 505155816 
	 while generalized-func is 505155816
